deepinv/models/swinir.py

```diff
@@ -0,0 +1,1111 @@
+            self.conv_last = nn.Conv2d(embed_dim, num_out_ch, 3, 1, 1)
+
+        self.apply(self._init_weights)
+
+        if pretrained is not None:
+            if pretrained == "download":
+                assert img_size == 128
+                assert in_chans in [1, 3]
+                assert upscale == 1
+                assert window_size == 8
+                assert img_range == 1.0
+                assert embed_dim == 180
+                assert mlp_ratio == 2
+                assert upsampler == ""
+                assert resi_connection == "1conv"
+                assert depths == [6, 6, 6, 6, 6, 6]
+                assert num_heads == [6, 6, 6, 6, 6, 6]
+
+                if in_chans == 1:
+                    weights_url = "https://github.com/JingyunLiang/SwinIR/releases/download/v0.0/004_grayDN_DFWB_s128w8_SwinIR-M_noise15.pth"
+                elif in_chans == 3:
+                    weights_url = "https://github.com/JingyunLiang/SwinIR/releases/download/v0.0/005_colorDN_DFWB_s128w8_SwinIR-M_noise15.pth"
+
+                pretrained_weights = torch.hub.load_state_dict_from_url(
+                    weights_url, map_location=lambda storage, loc: storage
+                )
+            else:
+                pretrained_weights = torch.load(
+                    pretrained, map_location=lambda storage, loc: storage
+                )
+            param_key_g = "params"
+            pretrained_weights = (
+                pretrained_weights[param_key_g]
+                if param_key_g in pretrained_weights.keys()
+                else pretrained_weights
+            )
+            self.load_state_dict(pretrained_weights, strict=True)
+
+    def _init_weights(self, m):
+        if isinstance(m, nn.Linear):
+            trunc_normal_(m.weight, std=0.02)
+            if isinstance(m, nn.Linear) and m.bias is not None:
+                nn.init.constant_(m.bias, 0)
+        elif isinstance(m, nn.LayerNorm):
+            nn.init.constant_(m.bias, 0)
+            nn.init.constant_(m.weight, 1.0)
+
+    @torch.jit.ignore
+    def no_weight_decay(self):
+        return {"absolute_pos_embed"}
+
+    @torch.jit.ignore
+    def no_weight_decay_keywords(self):
+        return {"relative_position_bias_table"}
+
+    def check_image_size(self, x):
+        _, _, h, w = x.size()
+        mod_pad_h = (self.window_size - h % self.window_size) % self.window_size
+        mod_pad_w = (self.window_size - w % self.window_size) % self.window_size
+        x = F.pad(x, (0, mod_pad_w, 0, mod_pad_h), "reflect")
+        return x
+
+    def forward_features(self, x):
+        x_size = (x.shape[2], x.shape[3])
+        x = self.patch_embed(x)
+        if self.ape:
+            x = x + self.absolute_pos_embed
+        x = self.pos_drop(x)
+
+        for layer in self.layers:
+            x = layer(x, x_size)
+
+        x = self.norm(x)  # B L C
+        x = self.patch_unembed(x, x_size)
+
+        return x
+
+    def forward(self, x, sigma=None):
```

Comment:
I would suggest to rename this forward

```
def forward_patch(self, x, sigma=None):
      ...
```

and to add a forward containing this small [snippet of code](https://github.com/JingyunLiang/SwinIR/blob/main/main_test_swinir.py#L68) (enabling to run on images of any size)

Reply:
What should be an equivalent snippet of code is already called at the beginning of the method `forward`.

https://github.com/deepinv/deepinv/blob/ac5ebe093a8d08575317058f93da22f78efbb418/deepinv/models/swinir.py#L1057